src/lib/transition_handler/processor.ml

```diff
@@ -11,43 +11,152 @@ open Async_kernel
+
+  (* add a breadcrumb and perform post processing *)
+  let add_and_finalize ~frontier ~catchup_scheduler
+      ~processed_transition_writer ~only_if_present cached_breadcrumb =
+    let breadcrumb =
+      if Cached.is_pure cached_breadcrumb then Cached.peek cached_breadcrumb
+      else Cached.invalidate_with_success cached_breadcrumb
+    in
+    let transition =
+      Transition_frontier.Breadcrumb.transition_with_hash breadcrumb
+    in
+    let add_breadcrumb =
+      if only_if_present then Transition_frontier.add_breadcrumb_if_present_exn
+      else Transition_frontier.add_breadcrumb_exn
+    in
+    let%map () = add_breadcrumb frontier breadcrumb in
+    Writer.write processed_transition_writer transition ;
+    Catchup_scheduler.notify catchup_scheduler
+      ~hash:(With_hash.hash transition)
+
+  let process_transition ~logger ~trust_system ~verifier ~frontier
+      ~catchup_scheduler ~processed_transition_writer
+      ~transition:cached_initially_validated_transition =
+    let enveloped_initially_validated_transition =
+      Cached.peek cached_initially_validated_transition
+    in
+    let sender =
+      Envelope.Incoming.sender enveloped_initially_validated_transition
+    in
+    let initially_validated_transition =
+      Envelope.Incoming.data enveloped_initially_validated_transition
+    in
+    let {With_hash.hash= transition_hash; data= transition}, _ =
+      initially_validated_transition
+    in
+    let metadata = [("state_hash", State_hash.to_yojson transition_hash)] in
+    Deferred.map ~f:(Fn.const ())
+      (let open Deferred.Result.Let_syntax in
+      let%bind mostly_validated_transition =
+        let open Deferred.Let_syntax in
+        match
+          Transition_frontier_validation.validate_frontier_dependencies ~logger
+            ~frontier initially_validated_transition
+        with
+        | Ok t ->
+            return (Ok t)
+        | Error `Not_selected_over_frontier_root ->
+            let%map () =
+              Trust_system.record_envelope_sender trust_system logger sender
+                ( Trust_system.Actions.Gossiped_invalid_transition
+                , Some
+                    ( "$state_hash was not selected of transition frontier root"
+                    , metadata ) )
+            in
+            Error ()
+        | Error `Already_in_frontier ->
+            Logger.warn logger ~module_:__MODULE__ ~location:__LOC__ ~metadata
+              "refusing to process $state_hash because is is already in the \
+               transition frontier" ;
+            return (Error ())
+        | Error `Parent_missing_from_frontier ->
+            Catchup_scheduler.watch catchup_scheduler
+              ~timeout_duration:catchup_timeout_duration
+              ~cached_transition:cached_initially_validated_transition ;
+            return (Error ())
+      in
+      (* TODO: only access parent in transition frontier once (already done in call to validate dependencies) *)
```

Comment:
can you open an issue for this?

Reply:
https://github.com/CodaProtocol/coda/issues/2485

add issue # to comment as well